3scale/porta
# Container-native builds powered by make/rake

**What this PR does / why we need it**:

This PR aims to do 2 things:

- Give developers the chance to run the tests on their own machine (through a docker-compose set up that orchestrates the necessary containers required for running the tests). It should be noted that one can do so, without needing to install any of this project's dependencies on the host system, since dependencies are only installed in the containers and in local volumes. This should lower the entry barrier for both external contributors, as well as new members on the team.
 - Move test commands to make targets, rather than including them in the circleci config directly, so that they can be reused in the dev environment. At the same time this makes it easier to move to a different CI, if for any reason the need comes up... All we would need to do is have the CI server spin up a set of containers and invoke the make targets in the same way they are invoked in circleci. This was not as easy to do when we needed to provision VMs - now we just need containers.

In more detail, here is what has changed:

## Open Source Builder Container Image

### Source
Switched to OSS builder image which does NOT contain source code, but rather expects that it will be checked out inside the container (in CI env) or provided as a container volume (in dev env).

### Dependencies
The installation of dependencies used to happen when the builder docker image was being **built**. And the builder image was being built on every CI run (in fact, in each of the 7 parallel runs).

I have moved here to a model where the builder image only includes system and _NOT_ application dependencies. The application dependencies are still installed as part of the CI build (we want this), but this change should make builds somewhat faster / more efficient, given we no longer have to build the rest of the docker image.

Caching of the applications dependencies was already in place, so the *speedup* will not be _dramatic_ (installing application dependencies was probably the part of the container builder image creation that took most time), but we are still better off not doing the rest of the image building twice.

To accommodate this, Makefiles have been reworked to:
* if a target is really up-to-date we don't reinvoke it. Operations like precompiling assets, and installing dependencies should really only run when there is a need to, as they add considerable time.
* Moved all logic from docker.sh and jenkins.sh to make targets, so that we can use it without needing the scripts.
* proxy_env no longer necessary, since we're dropping squid
* removed special features from docker-compose (`cap_add`, `privileged` mode and `dns`, etc.) to make it easier to run the set of containers on any external contributor's machine.

### Backing Services

Finally, the new image we are switching to does not run the backing services any more. The focus is to run backing services as separate containers (both through docker-compose, as well as in the CI environment (circleci)).

## Container-based development environment

Motivation for this is for external contributors / new team members to be able to run the test suite in their local environment without much friction or requirements for installation of additional software packages.

In order to achieve this, the Makefiles and some bash scripts have had to be reworked:
* `make` targets now have to cater to 2 environments: `dev` and `test`. The `Makefile` contains the common targets, while the `container.mk` and `docker-compose.mk` include the environment-specific details for `dev` and `test` envs respectively,
* docker.sh and jenkins.sh are on their way out. the logic that lives there is on the way to being moved to `make` targets
* Removed cache container since upgrading docker-compose file version allows us to use named volumes.

**Which issue(s) this PR fixes** 

Fixes #93 
Supercedes #242 

**Verification steps** 

Once this PR is complete, an external contributor should be able to run make test (or similar) that will run the same tests that would run on circleci, but instead run them on the contributor's machine.

Pipelines should pass on both jenkins and circleci.

**Special notes for your reviewer**:

Please keep in mind that a lot of the changes in this PR are taking into account that we currently need to support both Jenkins and CircleCI runs. Once a decision is made to move to one vs. the other, the logic that is found in the makefiles would be significantly simplified. 

Future work that will be based on this PR will include both: 
* further improving the developer experience for external contributors (by providing more refined targets catered to the more common tasks they will need in their dev env, powered by docker compose)
* simplifying the logic in makefiles that caters to different CI environments.
